src/features/syntaxDecorations.ts

```diff
@@ -32,18 +32,26 @@ const decorationTypes: { [type: string]: TextEditorDecorationType } = {
   }),
 };
 
-const decors: { [decorTypeName: string]: Range[] } = {};
+const decors: { [fileName: string]: { [decorTypeName: string]: Range[] } } = {};
 
 const regexToDecorationTypes: { [regexp: string]: string[] } = {
   // [[ref]]
   [refPattern]: ['gray', 'lightBlue', 'gray'],
 };
 
-export const getDecorations = (textEditor: TextEditor): { [decorTypeName: string]: Range[] } => {
+export const getDecorations = (
+  textEditor: TextEditor,
+  recompute: boolean = false,
+): { [decorTypeName: string]: Range[] } => {
   const doc = textEditor.document;
 
+  if (!recompute && doc.fileName in decors) {
+    return decors[doc.fileName];
+  }
+
+  decors[doc.fileName] = {};
   Object.keys(decorationTypes).forEach((decorTypeName) => {
-    decors[decorTypeName] = [];
+    decors[doc.fileName][decorTypeName] = [];
```

Comment:
`decors[doc.fileName]` should be extracted to a local mapping in order to avoid repeatedly indexing it.